// jeu-backend/managers/RequestManager.js
/*jshint esversion:6*/

class RequestManager {
    constructor(app, DBConnection) {
        this.app = app;
        this.DBConnection = DBConnection;
        
        this.app
            .get("/", this.getMainPage.bind(this))
            .get("/menu/", this.getMenu.bind(this))
            .get("/game/:mode", this.getGamePage.bind(this))
            .get("/profile/:id", this.getProfile.bind(this))
            .get("/options/:id", this.getOptionsPanel.bind(this));
                
                
    }
    
    getMainPage(req, res) {
        res.render("tpl-connect-panel.ejs"); 
    }
    
    getMenu(req, res) {
        res.render("main-menu.ejs");
    }
    
    getGamePage(req, res) {
        if(req.params.mode === "local") {
        res.render("game.ejs");      
    
        } else {
            res.render("game_online.ejs");
        }
    }
    
    getProfile(req, res) {
        this.DBConnection.getUserProfileInfo(req.params.id).then((info) => {
       
            res.render("profile.ejs", {"info": info}); 
    
        });
    }
    
    getOptionsPanel(req, res) { // Blocked for non-authenticated users (for now)
        console.log("On arrive aux options");
        let id = req.params.id;
        this.DBConnection.getOptions(id).then((options) => {
           //options = JSON.stringify(options);
            console.log(options);
           res.render("options.ejs", {"options": options, "machin": 1+1}); 
        });
    }

    
}

module.exports = RequestManager;
