// jeu-backend/managers/RequestManager.js
/*jshint esversion:6*/

"use strict"

class RequestManager {
    constructor(app, DBConnection) {
        this.app = app;
        this.DBConnection = DBConnection;

        this.app
            .get("/", this.getMainPage.bind(this))
            .get("/menu/", this.getMenu.bind(this))
            .get("/game/:mode", this.getGamePage.bind(this))
            .get("/profile/:id/", this.getProfile.bind(this))
            .get("/options/", this.getOptionsPanel.bind(this))
            .get("/getoptions/:id/", this.sendOptions.bind(this))
            .post("/sendoptions/", this.setOptions.bind(this))
            .post("/connect", this.connect.bind(this));


    }

    getMainPage(req, res) {
        res.render("tpl-connect-panel.ejs");
    }

    connect(req, res) {

        let userData = req.body;

        return this.DBConnection.getUserData(userData).then((data) => {
           if(data === "WRONG_PASSWORD") {
                res.send("WRONG_PASSWORD")
           } else {
                res.send(data);

           }

       });
    }

    getMenu(req, res) {
        res.render("main-menu.ejs");
    }

    getGamePage(req, res) {
        if(req.params.mode === "local") {
            res.render("game.ejs");

        } else {
            res.render("game_online.ejs");
        }
    }

    getProfile(req, res) {
        this.DBConnection.getUserProfileInfo(req.params.id).then((info) => {

            res.render("profile.ejs", {"info": info});

        });
    }

    getOptionsPanel(req, res) { // Blocked for non-authenticated users (for now)

        res.render("options.ejs");
    }

    sendOptions(req, res) {
        let id = req.params.id;
        if(id == -1) {
            res.send({"volume": .5, 'animations': 1});
        }
        this.DBConnection.getOptions(id).then((options) => {
            res.send(options);
        });
    }

    setOptions(req, res) {
        this.DBConnection.setOptions(req.body).then(() => {
            res.send("");
        });
    }




}

module.exports = RequestManager;
