// jeu-backend/joueur.js
/*jshint esversion:6*/

"use strict";

const Arme = require("./arme");
class Joueur {
    constructor(id, nom) {
        this.nom = nom;
        this.id = id;
        this.arme = new Arme(1);
        this.posture = false; // Attaque : true, défense : false
        this.pv = 100;
        this.position = [-1, -1],
        this.ancienne_arme = -1, // Sera utilisé en cas de ramassage d'une arme, pour la remettre à l'emplacement d'origine
        this.anc_position = [],
        this.tour_ramasse = -1;
    }

    updateArme(id_arme, nb_tours) {

        this.ancienne_arme = this.arme.id; // On met ici le niveau de l'ancienne arm
        this.tour_ramasse = nb_tours;
        this.arme = new Arme(id_arme); // On update l'objet arme
    }


    updatePosture(bool_posture) {
        this.posture = bool_posture;
    }

    recevoirDegats(arme_ennemi) {
        var degats = arme_ennemi.degats;

        if(!this.posture) {
            degats = Math.round(degats/2);
        }

        if(this.pv < degats) {
            degats = this.pv; // On définit le niveau de degâts au nombre de PV restants si l'ennemi va mourir ici pour ne pas avoir à le faire côté client (on lui enverra juste un signal différent)
            this.pv = 0;
        }

         else
            this.pv -= degats;

        return degats;
    }
}

module.exports = Joueur;
